src/lib/clean.c

```diff
@@ -57,61 +54,50 @@ static bool aa_hook_context_has_cache(AaHookContext *self, const char *d_name)
  */
 bool aa_hook_context_clean_cache(AaHookContext *self)
 {
-        autofree(DIR) *dir = NULL;
-        struct dirent *ent = NULL;
-        bool ret = false;
-        int fd = 0;
+        bool ret = true;
+        FTSENT *curr_entry = NULL;
+        FTS *file_system = NULL;
 
         if (aa_unlikely(!self)) {
                 return false;
         }
 
-        /* If the directory doesn't exist yet, it isn't fatal */
-        dir = opendir(self->cache_dir);
-        if (!dir) {
-                if (errno == ENOENT) {
-                        return true;
-                }
+        char *path[] = { self->cache_dir, NULL };
+        file_system = fts_open(path, FTS_PHYSICAL | FTS_NOSTAT, NULL);
+
+        if (file_system == NULL) {
                 return false;
         }
 
-        /* Assume success at this point */
-        fd = dirfd(dir);
-        ret = true;
-
-        while ((ent = readdir(dir)) != NULL) {
-                if (strlen(ent->d_name) == 1 && strcmp(ent->d_name, ".") == 0) {
-                        continue;
-                }
-
-                if (strlen(ent->d_name) == 2 && strcmp(ent->d_name, "..") == 0) {
-                        continue;
-                }
-
-                if (aa_hook_context_has_cache(self, ent->d_name)) {
+        while ((curr_entry = fts_read(file_system)) != NULL && errno == 0) {
+                unsigned short path_info = curr_entry->fts_info;
+                if (path_info != FTS_F || path_info != FTS_NSOK || path_info != FTS_DP) {
                         continue;
                 }
-
-                /* Don't wipe `.features` file. */
-                if (strcmp(ent->d_name, ".features") == 0) {
-                        continue;
+                if (path_info == FTS_F || path_info == FTS_NSOK) {
+                        /* Ignore the .features file and profiles that are still installed */
+                        if (strcmp(curr_entry->fts_name, ".features") == 0 ||
+                            aa_hook_context_has_cache(self, curr_entry->fts_name)) {
+                                continue;
+                        }
                 }
-
-                /* Continue even if the wipe fails so we clean some stuff up */
-                if (unlinkat(fd, ent->d_name, 0) != 0) {
-                        fprintf(stderr,
-                                "Unable to unlink() %s/%s: %s\n",
-                                self->cache_dir,
-                                ent->d_name,
-                                strerror(errno));
-                        ret = false;
-                        continue;
+                if (remove(curr_entry->fts_path) == 0) {
```

Comment:
I'll let this slide for now, but in the future, try to make the logic as similar as possible to the old code. If nothing else, it makes reviewing easier. I didn't see this last night, but you inverted this check compared to L101 of the old code. The previous method was actually cleaner, as it would have allowed you to combine the two if checks into:

```
if (remove(curr_entry->fts_path) != 0 && errno != ENOTEMPTY) {
    // print error and continue
}
//print success
```

Reply:
Your suggested solution is actually what I used in `while` condition. I'll edit it also for the sake of consistency.